packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -122,9 +140,35 @@ export class SchemaConverter implements SimpleDbSchemaConverter {
       p = p.next(() => this.ensureSequenceNumbers(txn));
     }
 
+    assert(
+      fromVersion <= 8,
+      'After version 8, we switch to a separately maintained schema ' +
+        'version. IndexedDB should never see its version go past 8.'
+    );
+    if (fromVersion < 8 && toVersion >= 8) {
+      p = p.next(() => addManualSchemaVersion(txn, db));
+    }
+
+    // IMPORTANT: new migrations should be implemented in
+    // `upgradeFromManualVersion`. After this point, we no longer use
+    // IndexedDB's versioning, as it does not allow downgrades.
+
     return p;
   }
 
+  doManualMigrations(
+    db: IDBDatabase,
+    toVersion: number
+  ): PersistencePromise<void> {
+    const objectStores = Array.from(db.objectStoreNames);
+    const txn = new SimpleDbTransaction(
+      db.transaction(objectStores, 'readwrite')
+    );
+    return readManualSchemaVersion(txn).next(fromVersion =>
+      this.upgradeFromManualVersion(txn, fromVersion, toVersion)
```

Comment:
I just realized:

1. upgradeFromManualVersion() will need a reference to `db` in order to create / delete object stores, etc.
2. But you can only call `db.createObjectStore()`, etc. from within the implicit "versionchange" transaction that's opened during the onupgradeneeded event...  but we're running migrations *after* the database has been opened and the "versionchange" transaction has completed. So any attempt to create object stores will fail.

So I think for this to work, we need to try to move the manual migrations into the onupgradeneeded handler...  Is that possible?  Or is it time to rethink web downgrade-ability?